landlab/ca/cfuncs.pyx

```diff
@@ -187,23 +178,23 @@ cdef int current_link_state(
 @cython.boundscheck(True)
 @cython.wraparound(False)
 cpdef update_link_states_and_transitions(
-    np.ndarray[DTYPE_INT_t, ndim=1] active_links,
-    np.ndarray[DTYPE_INT_t, ndim=1] node_state,
-    np.ndarray[DTYPE_INT_t, ndim=1] node_at_link_tail,
-    np.ndarray[DTYPE_INT_t, ndim=1] node_at_link_head,
-    np.ndarray[DTYPE_INT8_t, ndim=1] link_orientation,
-    np.ndarray[DTYPE_INT8_t, ndim=1] bnd_lnk,
-    np.ndarray[DTYPE_INT_t, ndim=1] link_state,
-    np.ndarray[DTYPE_INT_t, ndim=1] n_xn,
+    const id_t [:] active_links,
+    const id_t [:] node_state,
+    const id_t [:] node_at_link_tail,
+    const id_t [:] node_at_link_head,
+    const int8_t [:] link_orientation,
+    int8_t [:] bnd_lnk,
+    id_t [:] link_state,
+    id_t [:] n_xn,
     event_queue,
-    np.ndarray[DTYPE_t, ndim=1] next_update,
-    np.ndarray[DTYPE_INT_t, ndim=2] xn_to,
-    np.ndarray[DTYPE_t, ndim=2] xn_rate,
-    DTYPE_INT_t num_node_states,
-    DTYPE_INT_t num_node_states_sq,
-    DTYPE_t current_time,
-    np.ndarray[DTYPE_INT8_t, ndim=2] xn_propswap,
-    xn_prop_update_fn,
+    cython.floating [:] next_update,
+    id_t [:, :] xn_to,
+    cython.floating [:, :] xn_rate,
+    long num_node_states,
+    long num_node_states_sq,
+    double current_time,
+    int8_t [:, :] xn_propswap,
+    np.ndarray[object, ndim=2] xn_prop_update_fn,
```

Comment:
I don't recall why it's 2d. I'll have to double check if that's correct or not.

Reply:
@gregtucker It's 2D because it actually is a 2D array of functions. It's used in ``get_next_event`` like,
```python
prop_update_fn = xn_prop_update_fn[current_state, i]
```